### PR TITLE
feature: implement setTimeout, setInterval, clearTimeout, and clearInterval

### LINKED ISSUE #63: setTimeout support?
`setTimeout` does not seem to work

### PR DESCRIPTION
**Pull request by JakeChampion**
Using the new fastly_async_io module we can implement JS timers 🥳 

This pull-request implements setTimeout, setInterval, clearTimeout, and clearInterval and adds some interface and basic tests for them. There are also tests in the web-platform-tests that exercise the functions fully, which is why I only needed to write some basic tests.

Currently Viceroy does not have a host implementation for fastly_async_io so for now our tests can only run in production.

Closes #63

### CODE REVIEW

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3740]: What happens if the `timers` list is empty? If this function won't be called in that case, it would be worth adding a comment here explaining why a null check isn't necessary.

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3720]: This will read arbitrary memory if the list is empty. Is it only called in cases where you know the list is not empty?

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3778]:
```suggestion
    for (auto &arg : timer->arguments) {
      argv.infallibleAppend(arg);
    }
```

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3781]: Another option would be to put the code that access `timer` into a lexical block, so that it's impossible to name it after the point that it's consumed.

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3788]: In what situations would the timer at the head of the queue no longer be the one we just ran? Would it be easier to pop the timer off of the front of the queue and take ownership of it for the duration of the processing? That way you wouldn't have to worry about whether or not something was pushed in the meantime.

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:4476]:
```suggestion
  uint32_t id = timers->add_timer(handler, delay, handler_args, repeat);
```

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:4622]: Is it possible for a timer that repeatedly sets another timer with delay `0` to starve the loop that processes `pending_async_tasks`?

**Review by elliottt — COMMENTED** [review]
This is looking great! I haven't had to use `chrono` much, so this was fun to read through :D

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:4622]: I don't believe it would starve the event loop, I ran the below in Firefox's console and could run other functions after it:
```js
function starve() {
console.log(2)
setTimeout(starve, 0)
}
starve()
console.log(1)
```

I'll see if I can come up with a test C@E application for this

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp]: It was a duplicate of an earlier one, I've removed it 👍

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp]: Once I moved `Timer` and `ScheduledTimers` into a namespace I then found that this function was not used anywhere and so I've removed it - I did find this surprising though

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp]: Nice!

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3699]: good spot, I've removed the use of setting `deadline` via a parameter -- it was leftover from a refactor

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3777]: Good idea! I've moved the code into a lexical block now 👍

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3687]: They're now wrapped in a namespace 👍

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp]: It is only called when we know there is an item in the list, however I have also added an assert to this function and also changed it to use `.front()` instead:
```cpp
Timer *first() {
    MOZ_RELEASE_ASSERT(timers.size() > 0);
    return timers.front();
  }
```

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp]: Update: the one location where we did a comparison on Timers was updated to instead compare on timer.deadline, which is why this operator was not being used anywhere

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3791]: I believe this can happen if the timer being ran adds a new timer which has a deadline that is earlier than the timer being ran. I'll see if that is actually possible to occur within a JS application

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:4627]: This explains why setTimeout with a duration of 0 doesn't starve the loop -- https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#zero_delays

Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:4627]: You can starve the loop with Promises though :D
```js
function starveWithPromises() {
console.log(3)
Promise.resolve().then(starveWithPromises)
}
starveWithPromises()
console.log(9) // this will never run
```

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp]: I've changed this once more now, it now checks to see if the list is empty and if so, it returns nullptr, otherwise it returns the front of the list

**Comment by JakeChampion** [comment]
@elliottt this is ready for another rounds of reviews -- I also used the version of Viceroy which has implementations for the `fastly_async_io` module so that I could update our web-platform-test results -- it's so good to see them all passing ☺️

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3719]:
```suggestion
      return timers.front();
```

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3740]: A comment would be good here, thanks!

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [inline c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3742]: Does this work? If so, it might be nice to skip having to multiply by 1000 :)
```suggestion
    timer->deadline = system_clock::now() + std::milliseconds(timer->delay);
```

File: c-dependencies/js-compute-runtime/js-compute-builtins.cpp
Comment (elliottt) [reply c-dependencies/js-compute-runtime/js-compute-builtins.cpp:3794]: Ah I see: if this timer is set to repeat, then it moves the timer back in the queue rather than re-allocate the structure.

**Review by elliottt — APPROVED** [review]
This looks good to me!

**Comment by JakeChampion** [comment]
This is currently blocked on a new version of Viceroy which includes the `fastly_async_io` module -- https://github.com/fastly/Viceroy/pull/188

Once that pull-request is merged and released, we can update our version of Viceroy and should then see all the tests pass on this PR 👍

**Pull request merged**
